sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,133 @@
+- Decide to support a smaller number of backends for kubernetes officially if it can't be fixed.
+
+
+## Apiserver spending too much cpu/memory in duplicate Endpoints serializations
+
+#### Issue
+
+Currently while serving watches, the apiserver is deep-copying deserialized endpoints objects from etcd and serializing it once for each kube-proxy watch it serves (which is 5k watches in our largest clusters).
+
+#### Relevant issues/links
+
+- [#55779](https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-354452477)
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+- [#58050](https://github.com/kubernetes/kubernetes/issues/58050)
+
+#### Possible Solution(s)
+
+- Avoid those duplicate serializations if/when possible ([#60067](https://github.com/kubernetes/kubernetes/pull/60067) was an attempt to do this, but it didn't seem to help too much).
+- Moving from 'endpoints' -> 'endpoint' API change (discussed in detail in below issues) would mitigate this problem as the serialization load would decrease with smaller objects.
+
+
+## Endpoints traffic is quadratic in the number of endpoints
+
+#### Issue
+
+Endpoints object for a service contains all the individual endpoints of that service. As a result, whenever even a single pod in a service is added/updated/deleted, the whole endpoints object (which includes even the other endpoints that didn't change) is re-computed, written to storage and sent to all readers. Not being able to efficiently read/update individual endpoint changes can lead to (for e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If you consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+Overall, this is a serious performance drawback affecting multiple components in the control-plane (apiserver, etcd, endpoints-controller, kube-proxy). The current endpoints API (which was designed at a very early stage of kubernetes when people weren't really thinking too much about scalability/performance) makes it hard to solve this problem without introducing breaking changes.
+
+#### Relevant issues/links
+
+- [#47787](https://github.com/kubernetes/kubernetes/issues/47787)
+- [#24552](https://github.com/kubernetes/kubernetes/issues/24552)
+- [#24553](https://github.com/kubernetes/kubernetes/issues/24553)
+- [#8190](https://github.com/kubernetes/kubernetes/issues/8190)
+
+#### Possible Solution(s)
+
+- Redesign endpoints API with (at least) the following properties:
+  - Lives in an API group
+  - Not coupled to service resource lifecycle
+  - Can track both pod and non-pod endpoints
+  - Separate API object for each individual endpoint
```

Comment:
This is going to increase overhead a lot, so we may need spans or tranches.  Daemonset services are definitely the most likely to blow up first.

Reply:
> This is going to increase overhead a lot, so we may need spans or tranches. 

I think that's not clear to me. I'm not saying "this won't be needed for sure", but I think I'm not convinced it will for sure be needed.

> Daemonset services are definitely the most likely to blow up first.

I don't understand that. Can you clarify? What daemonset services? Kube-proxy e.g.? Why?

Reply:
Having one API object per endpoint is going to be more fixed overhead (serialization, key fetches, overhead of wrapping the object in memory) than having ranges of endpoints.

For reference, the metadata field itself is 200 bytes, even if every field is nil.  Maps can be up to 300 bytes in go, even empty, and metadata has 2 of them.

Having O(N) endpoints to fetch instead of O(1) endpoints with a small constant factor to fetch is going to be a tradeoff.

Reply:
> Having O(N) endpoints to fetch instead of O(1) endpoints with a small constant factor to fetch is going to be a tradeoff.

I agree with that. I was just trying to say that it's not obvious to me that what results of the experiment will be.

Reply:
And there are other optimizations we could make - e.g. not using map, if that is really a problem.

Reply:
> Having O(N) endpoints to fetch instead of O(1) endpoints with a small constant factor to fetch is going to be a tradeoff.

While I agree the small constant factor will affect fetching 'N' O(1) endpoints, we also need to consider how many times do we actually perform the operation of "get all endpoints" - as we seem to be losing only in such cases. With watches in use (which is the case at all key places FMU), we actually should be much better off getting only O(1) update for a single endpoint change instead of O(N) update. There's still the initial list call which suffers (but that should be very infrequent). Wdyt?